### PR TITLE
Preserve the bounds of the type parameters of used classes

### PR DESCRIPTION
**Pull request by kelloggm**
This should be part of the modularity model. The previous test case that showed we didn't do this was wrong, and I should not have approved it.

Should fix JDK-8319461.

### CODE REVIEW

**Comment by kelloggm** [comment]
If this passes CI, I will override approval requirements and merge it without review from @LoiNguyenCS. He should review it later when he gets a chance.

**Review by LoiNguyenCS — APPROVED** [review]
LGTM. 

I was aware of the fact that Specimin did not preserve a type bound if that type was not used by the target methods. I just assumed that removing the unused type bound should not affect an analysis tool.

**Comment by kelloggm** [comment]
> I just assumed that removing the unused type bound should not affect an analysis tool.

No worries, I should have noticed earlier. This is why we weren't preserving one of the JDK bugs, though, so I wanted to fix it.

**Comment by kelloggm** [comment]
@LoiNguyenCS next week you should look through the other changes I made in this PR to keep CF-6388 compiling. They're pretty extensive; we had the modularity model for interfaces wrong, too.

**Pull request merged**
